sklearn/preprocessing/data.py

```diff
@@ -1766,138 +1781,257 @@ class OneHotEncoder(BaseEstimator, TransformerMixin):
-
-        self.n_values_ = n_values
-        n_values = np.hstack([[0], n_values])
-        indices = np.cumsum(n_values)
-        self.feature_indices_ = indices
+            raise TypeError(error_msg)
 
-        column_indices = (X + indices[:-1]).ravel()
-        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
-                                n_features)
-        data = np.ones(n_samples * n_features)
-        out = sparse.coo_matrix((data, (row_indices, column_indices)),
-                                shape=(n_samples, indices[-1]),
-                                dtype=self.dtype).tocsr()
+    def _initialize_values(self):
+        """Standardize the `values` input
 
-        if (isinstance(self.n_values, six.string_types) and
-                self.n_values == 'auto'):
-            mask = np.array(out.sum(axis=0)).ravel() != 0
-            active_features = np.where(mask)[0]
-            out = out[:, active_features]
-            self.active_features_ = active_features
+        Output is either a string or a list of arrays.
+        """
+        if self.n_values is not None:
+            warnings.warn('`n_values` has been renamed to `values`.'
+                          'The parameter `n_values` has been deprecated '
+                          'and will be removed in version 0.21, use the'
+                          'parameter `values` instead and specify the '
+                          'expected values for each feature')
+            values = self.n_values
+        else:
+            values = self.values
 
-        return out if self.sparse else out.toarray()
+        # Convert `int` and `Sequence[int]` inputs to `List[Array[int]]`
+        if (not isinstance(values, six.string_types) and
+                np.isscalar(values)):
+            values = np.ones(self.n_features_cat_, dtype=int) * values
+        if (not isinstance(values, six.string_types) and
+                np.isscalar(values[0])):
+            values = [np.arange(v, dtype=np.int) for v in values]
+
+        return values
+
+    def _fit(self, X):
+        """Assumes `X` contains only categorical features"""
+        n_samples, n_features = X.shape
+        self.n_features_cat_ = n_features
+        self._label_encoders = [LabelEncoder() for i in range(n_features)]
+
+        self._values = self._initialize_values()
+        self._check_values(self._values, n_features)
+
+        # Fit on categorical features in the data
+        _auto_int_classes = n_features * [None]
+        for i in range(n_features):
+            le = self._label_encoders[i]
+
+            if np.isscalar(self._values) and self._values == 'auto':
+                # For integer features, allow integers between
+                # 0 and column max. The transform will still only
+                # return dummy columns for integers present in training data.
+                if (not isinstance(X[0, i], six.string_types) and
+                        int(X[0, i]) == X[0, i]):
```

Comment:
I think this is the weakest part of this PR. I'm type-checking the first row of the column and seeing if it looks like an integer, then if it is, assuming that the entire column is integers. I don't think we can avoid type-checking, since the encoder needs to treat integers differently than everything else in "auto" mode. I don't want to do type checking on every entry in the column, though. That seems too slow.

The `np.unique` call will error if the column has mixed string and numerical types, but we could still run into problems if there's mixed floats and integers.